Mathlib/Tactic/RCases.lean

```diff
@@ -0,0 +1,521 @@
+  hypothesis (which has the effect of replacing `b` with `a` everywhere or vice versa).
+* A type ascription `p : ty`, which sets the type of the hypothesis to `ty` and then matches it
+  against `p`. (Of course, `ty` must unify with the actual type of `h` for this to work.)
+* A tuple pattern `⟨p1, p2, p3⟩`, which matches a constructor with many arguments, or a series
+  of nested conjunctions or existentials. For example if the active hypothesis is `a ∧ b ∧ c`,
+  then the conjunction will be destructured, and `p1` will be matched against `a`, `p2` against `b`
+  and so on.
+* An alternation pattern `p1 | p2 | p3`, which matches an inductive type with multiple constructors,
+  or a nested disjunction like `a ∨ b ∨ c`.
+
+The patterns are fairly liberal about the exact shape of the constructors, and will insert
+additional alternation branches and tuple arguments if there are not enough arguments provided, and
+reuse the tail for further matches if there are too many arguments provided to alternation and
+tuple patterns.
+
+This file also contains the `obtain` and `rintro` tactics, which use the same syntax of `rcases`
+patterns but with a slightly different use case:
+
+* `rintro` (or `rintros`) is used like `rintro x ⟨y, z⟩` and is the same as `intros` followed by
+  `rcases` on the newly introduced arguments.
+* `obtain` is the same as `rcases` but with a syntax styled after `have` rather than `cases`.
+  `obtain ⟨hx, hy⟩ | hz := foo` is equivalent to `rcases foo with ⟨hx, hy⟩ | hz`. Unlike `rcases`,
+  `obtain` also allows one to omit `:= foo`, although a type must be provided in this case,
+  as in `obtain ⟨hx, hy⟩ | hz : a ∧ b ∨ c`, in which case it produces a subgoal for proving
+  `a ∧ b ∨ c` in addition to the subgoals `hx : a, hy : b |- goal` and `hz : c |- goal`.
+
+## Tags
+
+rcases, rintro, obtain, destructuring, cases, pattern matching, match
+-/
+
+namespace Lean.Parser.Tactic
+
+declare_syntax_cat rcasesPat
+syntax rcasesPatMed := sepBy1(rcasesPat, " | ")
+syntax rcasesPatLo := rcasesPatMed (" : " term)?
+syntax (name := rcasesPat.one) ident : rcasesPat
+syntax (name := rcasesPat.ignore) "_" : rcasesPat
+syntax (name := rcasesPat.clear) "-" : rcasesPat
+syntax (name := rcasesPat.tuple) "⟨" rcasesPatLo,* "⟩" : rcasesPat
+syntax (name := rcasesPat.paren) "(" rcasesPatLo ")" : rcasesPat
+
+declare_syntax_cat rintroPat
+syntax (name := rintroPat.one) rcasesPat : rintroPat
+syntax (name := rintroPat.binder) "(" rintroPat+ (" : " term)? ")" : rintroPat
+
+end Lean.Parser.Tactic
+
+/- A list, with a disjunctive meaning (like a list of inductive constructors, or subgoals) -/
+local notation "ListΣ" => List
+
+/- A list, with a conjunctive meaning (like a list of constructor arguments, or hypotheses) -/
+local notation "ListΠ" => List
+
+namespace Lean.Meta
+
+/-- Constructs a substitution consisting of `s` followed by `t`.
+  This satisfies `(s.append t).apply e = t.apply (s.apply e)` -/
```

Comment:
```suggestion
  This satisfies `(s.append t).apply e = t.apply (s.apply e)`
  for `e` that are in the domain of `s`. -/
```

AFAICT, most tactics only return substitutions for the variables that they changed (i.e. `subst x` would not return a substitution for an `α : Type` at the beginning of the local context).

Reply:
No, the original statement is correct. Note that `s.apply e` will return `e` unchanged if it is not in the substitution map. BTW, I think that any tactic that provides a `FVarSubst` should also take a `FVarSubst` as input, which serves no purpose other than to be the base for insertion. This is essentially the same as the `dlist` encoding of lists: it is more efficient to carry along a running substitution and add to it rather than using `append` (which is a workaround for those tactics that don't provide differential `FVarSubst` output).

Reply:
> Note that s.apply e will return e unchanged if it is not in the substitution map. 

This is precisely the issue.  If `s` doesn't contain `e` but `t` does, then `e` won't appear in the result of `append`.

Say you have the following context:
```
x y : Nat
h1 : x = y
z w : Nat
h2 : z = w
```
Then `subst h2` will return `s = {w ↦ w', h2 ↦ h2'}`.  If you call `subst h1` after that, you get `t = {y ↦ y', h1 ↦ h1'}`.  Note that `s.append t = s` keeps `y` unchanged even though it was renamed by the second subst.  (This is all based on just reading the code; I didn't run it.  I hope I'm wrong.)

Reply:
Let me put it another way: `(s.append t).apply e = t.apply (s.apply e)` is the *definition* of `append`, and the implementation is whatever is necessary to make this equation true. In your example `s` and `t` have disjoint domains, so `s.append t` should just be the union of the maps. As far as I can tell this is also the behavior of the implementation: it inserts every element in `s` to `t` (the `t.apply v` call just reduces to `v`). We get `{y ↦ y', h1 ↦ h1', k ↦ k'} |>.insert w w' |>.insert h2 h2' = {y ↦ y', h1 ↦ h1', k ↦ k', w ↦ w', h2 ↦ h2'}` as expected.

Reply:
Ah, nevermind.  You're right.  I missed the `t` at the end.